src/main/java/com/pi4j/crowpi/components/UltrasonicDistanceSensorComponent.java

```diff
@@ -1,15 +1,43 @@
 package com.pi4j.crowpi.components;
 
 import com.pi4j.context.Context;
+import com.pi4j.crowpi.components.events.SimpleEventHandler;
 import com.pi4j.crowpi.components.exceptions.MeasurementException;
 import com.pi4j.io.gpio.digital.*;
 
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
 
 /**
  * Implementation of the CrowPi ultrasonic distance sensor (HC-SR04) using GPIO with Pi4J
  */
 public class UltrasonicDistanceSensorComponent extends Component {
+    /**
+     * Scheduler instance for running the poller thread.
+     */
+    private final ScheduledExecutorService scheduler;
+    /**
+     * Active poller thread or null if currently not running.
+     */
+    private ScheduledFuture<?> poller;
+
+    /**
+     * Atomic simple event handler for "objectFound" and "objectDisappeared" event.
+     */
+    private final AtomicReference<SimpleEventHandler> objectFoundHandler;
+    private final AtomicReference<SimpleEventHandler> objectDisappearedHandler;
+    /**
+     * Variables used to generate SimpleEvents to provide object found and disappeared
+     */
+    private final AtomicBoolean state;
+    private double minRange;
+    private double maxRange;
+    private double temperature;
```

Comment:
One more thing I just though of: You should probably declare those `volatile` so that changes to these variables will immediately be visible to the poller thread.

Also while thinking about it, what do you think about adding setters/getters for these variables and them removing them from the `onXXX()` event handler methods? I just gave this a second thought and it's a little confusing that even the variables passed into `onObjectDisappeared` will overwrite the variables of `onObjectFound`. Like:

- `setDetectionRange(double min, double max)`
- `setMeasurementTemperature(double temperature)` - would also serve as the default value for `double measure()`
- `onObjectFound(SimpleEventHandler handler)`
- `onObjectDisappeared(SimpleEventHandler handler)`

Then one could also simply disable the event handlers by calling `onObjectFound(null)` or `onObjectDisappeared(null)` without having to specify zero values which are not necessary in the first place. And we could use a default temperature (already exists) and detection range to keep usage simple.